domain/resourcegroup/runaway.go

```diff
@@ -121,38 +205,90 @@ func (rm *RunawayManager) DeriveChecker(resourceGroupName string, originalSQL st
 	if group.RunawaySettings == nil {
 		return nil
 	}
-	return newRunawayChecker(rm, resourceGroupName, group.RunawaySettings, originalSQL, planDigest)
+	return newRunawayChecker(rm, resourceGroupName, group.RunawaySettings, originalSQL, sqlDigest, planDigest)
 }
 
-func (rm *RunawayManager) markQuarantine(resourceGroupName, convict, watchType string, ttl time.Duration, action string, now *time.Time) {
-	key := resourceGroupName + "/" + convict
-	if rm.watchList.Get(key) == nil {
-		rm.queryLock.Lock()
-		if rm.watchList.Get(key) == nil {
-			rm.watchList.Set(key, struct{}{}, ttl)
-		}
-		rm.queryLock.Unlock()
-	}
+func (rm *RunawayManager) markQuarantine(resourceGroupName, convict string, watchType rmpb.RunawayWatchType, action rmpb.RunawayAction, ttl time.Duration, now *time.Time) {
 	var endTime time.Time
 	if ttl == 0 {
-		endTime = longestTime
+		endTime = LongestTime
 	} else {
 		endTime = now.Add(ttl)
 	}
-	select {
-	case rm.quarantineChan <- &QuarantineRecord{
+	record := &QuarantineRecord{
 		ResourceGroupName: resourceGroupName,
 		StartTime:         *now,
 		EndTime:           endTime,
 		Watch:             watchType,
 		WatchText:         convict,
 		Source:            rm.serverID,
-	}:
+		Action:            action,
+	}
+	rm.addWatchList(record, ttl, false)
+	select {
+	case rm.quarantineChan <- record:
 	default:
 		// TODO: add warning for discard flush records
 	}
 }
+
+func (rm *RunawayManager) addWatchList(record *QuarantineRecord, ttl time.Duration, force bool) {
+	key := record.GetRecordKey()
+	if force {
+		rm.queryLock.Lock()
+		if rm.watchList.Get(key) != nil {
+			rm.watchList.Delete(key)
+		}
+		rm.watchList.Set(key, record, ttl)
+		rm.queryLock.Unlock()
+	} else {
+		if rm.watchList.Get(key) == nil {
+			rm.queryLock.Lock()
+			if rm.watchList.Get(key) == nil {
+				rm.watchList.Set(key, record, ttl)
+			} else {
+				rm.staleQuarantineRecord <- record
+			}
+			rm.queryLock.Unlock()
+		} else {
+			rm.staleQuarantineRecord <- record
+		}
+	}
+}
+
+func (rm *RunawayManager) GetWatchInWatchList(key string) *QuarantineRecord {
+	item := rm.watchList.Get(key)
+	if item == nil {
+		return nil
+	}
+	return item.Value()
+}
+
+func (rm *RunawayManager) AddWatch(record *QuarantineRecord) {
+	ttl := time.Until(record.EndTime)
+	if ttl <= 0 {
+		rm.staleQuarantineRecord <- record
```

Comment:
Why need this explicit stale. Outdated record should be auto cleaned up by the clean up loop

Reply:
I remove the the clean up loop. The reason for this is that the information_schema table was not involved before, and I wanted to gc it as quickly as possible